### PR TITLE
Add MeshCoP State TLV

### PR DESCRIPTION
**Pull request by luwannest**
MeshCoP State TLV is not optional MGMT_ACTIVE_SET.rsp and MGMT_PENDING_SET.rsp.

### CODE REVIEW

File: src/core/thread/meshcop_tlvs.hpp [hunk 1]
```diff
@@ -659,6 +660,59 @@ class ActiveTimestampTlv : public Tlv, public Timestamp
 } OT_TOOL_PACKED_END;
 
 /**
+ * This class implements State TLV generation and parsing.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+class StateTlv: public Tlv
```
Comment (jwhui) [inline src/core/thread/meshcop_tlvs.hpp]: You should borrow the implementation from my [pr/commissioning](https://github.com/jwhui/openthread/blob/pr/commissioning/src/core/thread/meshcop_tlvs.hpp#L936) branch.  See further comments below.

File: src/core/thread/meshcop_tlvs.hpp [hunk 2]
```diff
@@ -659,6 +660,59 @@ class ActiveTimestampTlv : public Tlv, public Timestamp
...
+{
+public:
+    /**
+      * State TLV values.
+      */
+    enum State
+    {
+        kAccept = 1,
```
Comment (jwhui) [inline src/core/thread/meshcop_tlvs.hpp]: nit: align `=` for the states.

File: src/core/thread/meshcop_tlvs.hpp [hunk 3]
```diff
@@ -659,6 +660,59 @@ class ActiveTimestampTlv : public Tlv, public Timestamp
...
+     */
+    bool IsValid(void) const { return GetLength() == sizeof(*this) - sizeof(Tlv); }
+
+    /**
+     * This method returns the State value.
+     *
+     * @returns The State value.
+     *
+     */
+    int8_t GetState(void) const { return mState; }
```
Comment (jwhui) [inline src/core/thread/meshcop_tlvs.hpp]: Return `State` type, rather than `int8_t`.

File: src/core/thread/meshcop_tlvs.hpp [hunk 4]
```diff
@@ -659,6 +660,59 @@ class ActiveTimestampTlv : public Tlv, public Timestamp
...
+
+    /**
+     * This method sets the State value.
+     *
+     * @param[in]  aState  The State value.
+     *
+     */
+    void SetState(int8_t aState) { mState = aState; }
```
Comment (jwhui) [inline src/core/thread/meshcop_tlvs.hpp]: Argument should have `State` type, rather than `int8_t`.

**Comment by jwhui** [comment]
Thanks for addressing this.  See comments inline.

File: src/core/thread/meshcop_tlvs.hpp (hunk 1 above)
Comment (luwannest) [inline src/core/thread/meshcop_tlvs.hpp]: Yes, I should dive deeper into the code. Thanks for the reminds!

**Comment by jwhui** [comment]
LGTM 👍

**Pull request merged**